src/tools/chip-cert/CertUtils.cpp

```diff
@@ -0,0 +1,826 @@
+        OPENSSL_free(name);
+    }
+    if (header != nullptr)
+    {
+        OPENSSL_free(header);
+    }
+    if (data != nullptr)
+    {
+        OPENSSL_free(data);
+    }
+
+    return res;
+}
+
+static CertFormat DetectCertFormat(uint8_t * cert, uint32_t certLen)
+{
+    static const uint8_t chipRawPrefix[]   = { 0xD5, 0x00, 0x00, 0x08, 0x00, 0x01, 0x00 };
+    static const char * chipB64Prefix      = "1QAACAAB";
+    static const uint32_t chipB64PrefixLen = sizeof(chipB64Prefix) - 1;
+    static const char * pemMarker          = "-----BEGIN CERTIFICATE-----";
+
+    if (certLen > sizeof(chipRawPrefix) && memcmp(cert, chipRawPrefix, sizeof(chipRawPrefix)) == 0)
+    {
+        return kCertFormat_Chip_Raw;
+    }
+
+    if (certLen > chipB64PrefixLen && memcmp(cert, chipB64Prefix, chipB64PrefixLen) == 0)
+    {
+        return kCertFormat_Chip_Base64;
+    }
+
+    if (ContainsPEMMarker(pemMarker, cert, certLen))
+    {
+        return kCertFormat_X509_PEM;
+    }
+
+    return kCertFormat_X509_DER;
+}
+
+bool ReadCert(const char * fileName, X509 *& cert)
+{
+    CertFormat origCertFmt;
+    return ReadCert(fileName, cert, origCertFmt);
+}
+
+bool ReadCert(const char * fileName, X509 *& cert, CertFormat & origCertFmt)
+{
+    bool res = true;
+    CHIP_ERROR err;
+    uint8_t * certBuf = nullptr;
+    const uint8_t * p;
+    uint32_t certLen;
+    CertFormat certFmt;
+
+    cert = nullptr;
+
+    res = ReadFileIntoMem(fileName, certBuf, certLen);
+    VerifyTrueOrExit(res);
+
+    VerifyOrExit(certLen <= MAX_CERT_SIZE, res = false);
+
+    certFmt = origCertFmt = DetectCertFormat(certBuf, certLen);
+
+    if (certFmt == kCertFormat_X509_PEM)
+    {
+        res = X509PEMToDER(certBuf, certLen);
+        VerifyTrueOrExit(res);
+
+        certFmt = kCertFormat_X509_DER;
+    }
+    else if (certFmt == kCertFormat_Chip_Base64)
+    {
+        if (!Base64Decode(certBuf, certLen, certBuf, certLen, certLen))
+        {
+            ExitNow(res = false);
+        }
+        certFmt = kCertFormat_Chip_Raw;
+    }
+
+    if (certFmt == kCertFormat_Chip_Raw)
+    {
+        uint32_t convertedCertLen  = certLen * kMaxChipCertInflationFactor;
+        uint8_t * convertedCertBuf = reinterpret_cast<uint8_t *>(malloc(convertedCertLen));
+
+        err = ConvertChipCertToX509Cert(certBuf, certLen, convertedCertBuf, convertedCertLen, convertedCertLen);
+        if (err != CHIP_NO_ERROR)
+        {
+            free(convertedCertBuf);
+            fprintf(stderr, "Error converting certificate: %s\n", chip::ErrorStr(err));
+            ExitNow(res = false);
+        }
+
+        free(certBuf);
+        certBuf = convertedCertBuf;
+        certLen = convertedCertLen;
+
+        certFmt = kCertFormat_X509_DER;
+    }
+
+    p    = certBuf;
+    cert = d2i_X509(nullptr, &p, certLen);
+    if (cert == nullptr)
+    {
+        ReportOpenSSLErrorAndExit("d2i_X509", res = false);
+    }
+
+exit:
+    if (cert != nullptr && !res)
+    {
+        X509_free(cert);
```

Comment:
Can you use `std::unique_ptr<X509, X509FreeDeleter>` where `X509FreeDeleter` is a custom deleter that calls `X509_free(ptr)` ?

Reply:
done